src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+// Written in the D programming language.
+/**
+This module provides a composable reference count implementation in the form
+of `__RefCount`.
+
+License: $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Eduard Staniloiu
+
+Source: $(DRUNTIMESRC core/experimental/refcount.d)
+*/
+module core.experimental.refcount;
+
+///
+unittest
+{
+    auto rc = __RefCount.make!__RefCount();
+    assert(rc.isUnique);
+    {
+        auto rc2 = rc; // Construct rc2 by copy construction
+        assert(!rc.isUnique);
+        auto rc3 = __RefCount.make!__RefCount();
+        rc2 = rc3; // Assign rc3 into rc2; rc's ref count drops
+        assert(rc.isUnique);
+        rc3 = rc; // Assign rc into rc3; rc's ref count increases
+        assert(!rc.isUnique);
+        // rc2 and rc3 go out of scope here
+        // rc2 is the last ref to the `__RefCount` created in this block -> gets freed
+    }
+    assert(rc.isUnique);
+}
+
+/**
+A qualified reference counted `struct` that is intended to be composed by
+user-defined types that desire to implement manual memory management by means of
+reference counting.
+
+`__RefCount` was designed to be composed as a field inside a user-defined type.
+The user is responsible to initialize the `__RefCount` in that type's constructors.
+The user will call the `isUnique()` method to decide if this is the last
+reference to the enclosing type so memory can be safely deallocated.
+
+Implementation: The internal implementation of `__RefCount` uses `malloc`/`free`.
+
+Important: The `__RefCount` member must be initialized through a call to its
+constructor before being used.
+*/
+struct __RefCount
+{
+    import core.atomic : atomicOp;
+
+    alias CounterType = uint;
+    private shared CounterType* rc = null;
+
+    /*
+    Perform `rc op val` operation. Always use atomics as the counter is shared.
+
+    Returns:
+         The result of the operation.
+    */
+    @nogc nothrow pure @trusted scope
+    private CounterType rcOp(this Q, string op)(CounterType val) const
+    {
+        return cast(CounterType)(atomicOp!op(*(cast(shared CounterType*) rc), val));
+    }
+
+    /**
+    Creates a new `__RefCount` instance. Its memory is internally managed with
+    malloc/free.
+
+    Params:
+         _ = an unused int value; required because structs don't have a
+             user defined default constructor
+    */
+    @nogc nothrow pure @trusted scope
+    this(this Q)(int _)
+    {
+        // We are required to always use a shared support as the result of a `pure`
+        // function is implicitly convertible to `immutable`.
+
+        shared CounterType* support = cast(shared CounterType*) pureAllocate(CounterType.sizeof);
+        *support = 1; // Start with 1 to avoid calling an `addRef` from 0 to 1
```

Comment:
This will fail to compile with the new (not yet implemented) restricted shared, please do `*cast()support = 1;.

Reply:
I changed `shared CounterType*` to `shared(CounterType)*`, as it's the counter that should be `shared`, not the pointer to it (as @thewilsonator pointed out in private). This makes this line compile with `-preview=restrictiveshared` (https://github.com/dlang/dmd/pull/10142), but that seems like a bug in the restrictive shared implementation to me?

Reply:
I think so, probably best to change it anyway.

Reply:
> This makes this line compile with -preview=restrictiveshared 

No it doesn't... shared(int) can't be incremented.

Reply:
Perhaps `*cast(CounterType*)support = 1`?

Reply:
We should really have `cast(ref CounterType)lvalue` to perform lvalue casting...